src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
The first thing I tried is updating an empty Model since it require the least amount of code and almost everything can be kept as they are. But then the default constructor give the empty Model the timestamp as current time and nothing is updated. I considered explicitly assign the timestamp to some time in the past but it feels too much like a hack. Then I also want to resuse all the downloading new repo goodies like the progress bar without changing so much code so user knows why it takes so long in this case.

Having multiple exceptionally then a `thenCompose` doesn't help (like in the 1st code snippet, I tried a similar thing also) since according to the docs, if any CompletionStage got exception then every later stage also completes exceptionally (like in a unix pipe). I was trying to use `isCompleteExceptionally()` but then I didn't get anything out of it either.

I have also tried all other variants of exception handling in `CompletableFuture`. They are essentially buffed up `exceptionally` with additional info passed in but exactly the same restricting expressiveness so they don't help either.

The side-effects I talked about won't happen since we have this *pendingQueue* thing. But suppose we didn't have that one and tried to download same repos concurrently then the side-effects will become obvious.

So something that looks so simple turns out to be so annoying :P